packages/compass-sidebar/src/components/multiple-connections/active-connections/active-connection-navigation.tsx

```diff
@@ -60,42 +89,162 @@ const activeConnectionCountStyles = css({
   marginLeft: spacing[100],
 });
 
+const filterConnections = (
+  connections: Connection[],
+  regex: RegExp
+): FilteredConnection[] => {
+  const results: FilteredConnection[] = [];
+  for (const connection of connections) {
+    const isMatch = regex.test(connection.name);
+    const childMatches = filterDatabases(connection.databases, regex);
+
+    if (isMatch || childMatches.length) {
+      results.push({
+        ...connection,
+        isMatch,
+        databases: childMatches.length ? childMatches : connection.databases,
+        // databases: (!isMatch && childMatches.length) ? childMatches : connection.databases, // TODO: check with Ben
+      });
+    }
+  }
+  return results;
+};
+
+/**
+ *
+ */
+const filterDatabases = (
+  databases: Database[],
+  regex: RegExp
+): FilteredDatabase[] => {
+  const results: FilteredDatabase[] = [];
+  for (const db of databases) {
+    const isMatch = regex.test(db.name);
+    const childMatches = filterCollections(db.collections, regex);
+
+    if (isMatch || childMatches.length) {
+      results.push({
+        ...db,
+        isMatch,
+        collections: childMatches.length ? childMatches : db.collections,
+        // collections: (!isMatch && childMatches.length) ? childMatches : db.collections, // TODO: check with Ben
+      });
+    }
+  }
+  return results;
+};
+
+const filterCollections = (
+  collections: Collection[],
+  regex: RegExp
+): FilteredCollection[] => {
+  return collections
+    .filter(({ name }) => regex.test(name))
+    .map((collection) => ({ ...collection, isMatch: true }));
+};
+
+/**
+ * Take the starting expandedConnections, and add 'tempExpanded' to collapsed items that:
+ * - are included in the filterResults
+ * - they either are a direct match, or their children are a direct match
+ */
+const applyTempExpanded = (
```

Comment:
Feels like we're unnecessarily overcomplicating the logic here (I'm leaving a comment on this method, but this touches most of the code here) by mixing separate derived states into one.

Your source of truth for this is user triggered collapsed state (like literally clicking / pressing a button), filter regex, current active tab. From this you can always derive expanded state as `userExpanded || (filter && matchesFilter) || matchesActiveTab`

Reply:
I'm not sure if I follow what exactly you're proposing. I tried to have the "default + user controlled expanded" as a persisted state and the "filter expanded" as a derived memo. But then I had trouble covering 'User can still collapse these parents (while the filter is active).'
Basically the filter overrides the previous default + user decided state, but subsequent user actions override the filter-induced state.

Reply:
Having both of them in the state, they can override each other, otherwise I'm not sure how to support this

Reply:
I didn't mean redux specifically, I just meant that they need to be in the state as opposed to e.g. one of them being derived
I like the useReducer approach and will try if this would make the code more readable (if not simpler). 
let me explain the "tempExpanded": once we have them both in a single state, we run into another problem - how to persist the user choices (from before and during filtering)? my solution was to add a third state - "tempExpanded", which gets reverted afterwards.
I can definitely see how the code and the whole solution is not very easy to wrap one's head around. there are several factors which contribute to the complexity (the 3 levels, different default state of connections and databases and the "lasting default + user-choice" vs "temporary filter-induced" expanding).
I'll try to play around and see if I can make the code clearer.

Reply:
we could avoid "tempExpanded" if we're okay with just reverting to the "pre-filter" expanded state after the filter ended. then we could just make a copy and return to that - not perform a cleanup. it is not the current behaviour, but it seems good enough to me and it would make things simpler. I will try to find some other examples in the world and see what is the standard.

Reply:
User expanded and user filter input should be in the state, yes, they are state that we need to control, we are not in diagreement here. But the final filtered list and the final expanded state are derived values based on user expanded and user filter and the list itself.

We can avoid it by deriving the value and it would preserve the user input before, after, and during using the filter. I'm not sure if you had a chance to look at my example, but it literally does it without keeping any extra state for the expanded that is not the user input. Obviously the tree I'm using there is simpler, I understand that, but there is really no difference in the logic no matter how much nesting you'll introduce, it mostly changes how exactly you check that children are matching the filter.

Deriving this state by resetting the user expanded choice in some cases is also pretty close to [what was already happening](https://github.com/mongodb-js/compass/blob/0d53b866e82b3d3fcd96f78af88c4cc0a872e8c9/packages/compass-sidebar/src/modules/databases.ts#L107-L117) in the app, but removed in this PR

Reply:
I did have a look at your example, I think however that you've changed it since then - now it's not resetting everything to collapsed after the filter is cleared.

I am already moving things around with the reducer, having the actions listed like that will surely be more readable than having stuff happening in the effects. Afterwards if it still looks confusing, I will get rid of the tempExpanded.